### ISSUE TITLE
Custom 404 pages in localized sites

### ISSUE THREAD

**Issue by Sporiff** [withastro/astro#12175]
### Astro Info

```block
Astro                    v4.15.11
Node                     v22.9.0
System                   macOS (arm64)
Package Manager          npm
Output                   static
Adapter                  none
Integrations             auto-import
                         @astrojs/react
                         astro-expressive-code
                         @astrojs/mdx
                         @astrojs/tailwind
                         @astrojs/sitemap
                         @astrojs/markdoc
```

### Describe the Bug

I'm using Astro's i18n options to create a localized site in 4 languages: en, ja, zh, ko. A hard requirement for my use case is that we use the `prefixDefaultLocale: true` setting to ensure that the default locale (en) always appears in the path just like the others.

One thing I need is the ability to create a custom 404 page that gives users instructions on what to do next when they reach a page that doesn't exist. Before I updated to the multi-folder structure, this was achievable by placing a `404.astro` in the `pages` directory. However, this page was not localized properly due to there not being any locale detection present. Since moving to the new folder-based hierarchy, I noticed that the `404.astro` doesn't appear in the `dist` folder after running `astro build`.

Adding a custom `404.astro` to `pages/[lang]` works if you visit the slug directly (e.g. `/en/404`), but it doesn't work when actually hitting a 404. I've managed to override this in Vercel by putting in a Vercel redirect that takes any 404 to the `/en/404` route then using some client side scripting to push the user to the correct page.

### What's the expected result?

I would like to be able to create a custom `404.astro` in `pages` that routes to the custom `404.astro` in `pages/[lang]` so that localized 404 pages may be served. Ideally, it would be good to control this via `astro.config.mjs`. Having the ability to simply set up locale-based routing to the major pages like `index.astro` and `404.astro` would save a lot of time.

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/astro-no-localized-404

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by Southpaw1496** [comment]
I don't think this is something that it's possible for Astro to do. Because a 404 could happen on any invalid path, you can't handle it with file-based routing. To get around this, many static hosts simply look for a `404.html` page in the root of the site and serve that with all 404 responses, regardless of their path.

If you really wanted, you could make a middleware (which will execute on every request) that would check if the slug exists, and serve the correct 404 page based on the user's language. This would, however, require you to use `output: server` and mean that every request to your site would count as a function invocation on your hosting platform.

**Comment by Sporiff** [comment]
> I don't think this is something that it's possible for Astro to do. Because a 404 could happen on any invalid path, you can't handle it with file-based routing. To get around this, many static hosts simply look for a `404.html` page in the root of the site and serve that with all 404 responses, regardless of their path. So, when you make `src/pages/404.astro`, that renders to `/404.html` in the output site and is served by the static host.

This is the issue to be solved, I think. I understand that Astro can't know at build time what language someone is going to be using when they hit a nonexistent route (that makes no sense), but currently putting a `404.astro` in `src/pages` when using the `prefixDefaultLocale: true` setting fails to add the 404 file to the build output.

Of course, you can work around this by adding it to `public`, but this isn't a great workaround and isn't documented as far as I know.

**Comment by ematipico** [comment]
This is fixed

**Comment by agsoto** [comment]
I'm reopening this because I've attempted to achieve the exact same without success. I saw there was a PR addressing this issue, but it seems it was never merged. I tested the minimal reproducible example with the latest versions, and I still can't find reliable information on how to localize the 404 page. Has anyone found a workaround for this?

**Comment by florian-lefebvre** [comment]
Yeah I'm not exactly sure why this was closed. @ematipico do you have some context?

**Comment by leonedev** [comment]
I am not able to localize it too. Anyone has news about this?

**Comment by ematipico** [comment]
Yeah I don't remember why, unfortunately

**Comment by Maofree** [comment]
Hi
I have the same bug here
https://github.com/withastro/astro/issues/15299
with astrojs 5.16.15

I see a fix but I don't think it's been applied

**Comment by ArnovanDoesburg** [comment]
In case anyone is still struggling with translated 404 pages without enabling server mode, here’s a solution that worked for me:  

1. Render all translations on the 404 page  
   - I included all available language versions of the 404 content in the page. Each translation is wrapped in its own container or layout.  

2. Use a small inline client-side script
   - The script checks the URL to detect the language.  
   - Once the language is determined, it removes all other language containers, leaving only the relevant translation visible.  

This way, you can serve a translated 404 page for any language without switching Astro to server mode. It’s not the most elegant in terms of initial page size, but it’s a reliable workaround until native language-based 404 routing is fully supported.

https://retrobright.org/123 - English 404 (default)
https://retrobright.org/es/123 - Spanish 404
https://retrobright.org/jp/123 - English 404 (default language fallback for non-supported languages)

**Comment by ericswpark** [comment]
@ArnovanDoesburg I did something like this, but the client-side script would check the 404 URL and see if it had a language prefix (e.g. `/blog/unknown-url/` would be the main language and `/ko/blog/unknown-url/` would be Korean), and if so, redirect to the localized 404 page with a search query parameter that encodes the original URL. Then the localized page would get the original URL and use it for reporting purposes.

Would still be nice if this is natively supported by Astro.

**Comment by r0hitm** [comment]
I encountered this today as well. I am adding japanese localization to my site, and have a rewrite fallback setup. For the 404 page, I'm hitting `NoMatchingStaticPathFound` error for localized 404 page on a dynamic route: `src/pages/tags/[tag]/index.astro`, during dev. And during build it renders the english 404

**Comment by matthewp** [comment]
@astrobot-houston triage
